KAFKA-5697: prevent poll() from blocking forever

vvcephei opened the pull request:
Add the new stricter-timeout version of `poll` proposed in KIP-266.

The pre-existing variant `poll(long timeout)` would block indefinitely for metadata
updates if they were needed, then it would issue a fetch and poll for `timeout` ms 
for new records. The initial indefinite metadata block caused applications to become
stuck when the brokers became unavailable. The existence of the timeout parameter
made the indefinite block especially unintuitive.

This PR adds `poll(Duration timeout)` with the semantics:
1. iff a metadata update is needed:
    1. send (asynchronous) metadata requests
    2. poll for metadata responses (counts against timeout)
        - if no response within timeout, **return an empty collection immediately**
2. if there is fetch data available, **return it immediately**
3. if there is no fetch request in flight, send fetch requests
4. poll for fetch responses (counts against timeout)
    - if no response within timeout, **return an empty collection** (leaving async fetch request for the next poll)
    - if we get a response, **return the response**

The old method, `poll(long timeout)` is deprecated, but we do not change its semantics, so it remains:
1. iff a metadata update is needed:
    1. send (asynchronous) metadata requests
    2. poll for metadata responses *indefinitely until we get it*
2. if there is fetch data available, **return it immediately**
3. if there is no fetch request in flight, send fetch requests
4. poll for fetch responses (counts against timeout)
    - if no response within timeout, **return an empty collection** (leaving async fetch request for the next poll)
    - if we get a response, **return the response**

One notable usage is prohibited by the new `poll`: previously, you could call `poll(0)` to block for metadata updates, for example to initialize the client, supposedly without fetching records. Note, though, that this behavior is not according to any contract, and there is no guarantee that `poll(0)` won't return records the first time it's called. Therefore, it has always been unsafe to ignore the response.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
KAFKA-5697: prevent poll() from blocking forever

Several tests depended on being able to send a timeout of 0, but still
have the coordinator poll take non-zero time to do its work. I updated
them to send a long enough timeout for the coordinator to to the
required work.

vvcephei commented:
@guozhangwang I think this change is a good solution to the problem you reported in https://issues.apache.org/jira/browse/KAFKA-5697. WDYT?

I've tried tomake the change in a 'private' fashion so as not to 
change any public APIs, but let me know if you think it still needs a KIP.

Who else should review a change like this?

hachikuji commented:
Thanks for the patch. I knew we should be pretty close to being able to do this by now after some of the recent work. Note, however, that `ConsumerCoordinator.refreshCommittedOffsetsIfNeeded` still blocks.

The main thing to be careful of is probably the case you mentioned where users have depended on `poll(0)` in order to ensure they have a valid assignment. I think this is the only thing that has kept me from finishing this. We could do a KIP to be on the safe side, but I'm on the fence whether it's actually needed. Maybe the question is whether we should give users an alternative API to await an assignment.

guozhangwang commented:
I'll leave to @hachikuji and @mjsax to review and unblock you.

vvcephei commented:
Hey @hachikuji, Thanks for taking a look!

I'll add a timeout to `refreshCommittedOffsetsIfNeeded` as well, and also look at the failing integration tests.

I was unaware that folks would use `poll(0)` to wait for an assignment without fetching. I was thinking it was just to "drain" the consumer of previously fetched records. Although, now that you mention it, that should have been obvious from the tests I changed.

In that light, I do think I should go ahead and do a KIP, since with this change there would be no way to wait for an assignment without polling. That will also be a good way to seek feedback on whether or not we need the drain-previously-fetched use case, since that will also no longer be possible. And of course, users may need to increase their timeout value since the metadata updates will now count against it.

If it's not too distracting to you, I'll keep this PR open to sketch out the new method and to update the tests and work though any failures while I write the KIP.

hachikuji commented:
@vvcephei Yeah, that sounds good to me. I'm not sure how common that usage is--it's definitely convenient in testing--but it's probably a good idea to get some more visibility for the change to be on the safe side. Thanks for picking this up, by the way!

vvcephei committed:
add timeouts to coordinator without changing semantics

vvcephei committed:
remove unintentional synchronization from tests

vvcephei committed:
roll back changes to KafkaConsumerTest

vvcephei committed:
remove unintentional synchronization

vvcephei commented:
Just to record my current state on this...

I decided a good way to approach this was to lay the groundwork by introducing timeouts to every call in the `poll` path that *can* block, but not changing any semantics presented by the `KafkaConsumer` itself (by passing `Long.MAX_VALUE` to all the new internal timeout parameters).

I found in testing that altering the semantics of `poll`'s timeouts was very disruptive to our unit and integration tests both, hence my taking the conservative approach for now.

Of course, the new timeouts are of no value unless they are exposed in some way, and that's what I'm going to work on tomorrow. Once I have poked around enough to have a decent proposal, then I'll write the promised KIP.

vvcephei committed:
decrease metadata timeout from maxLong to 1 sec

vvcephei committed:
add unexpected thread names to verifyNoUnexpectedThreads

vvcephei committed:
testing 30s as the timeout for metadata updates

vvcephei committed:
test in progress

vvcephei committed:
KafkaConsumerTest passes with async meta

vvcephei commented:
rebased

vvcephei committed:
(lower) bound timeout at 0 for ConsumerNetworkClient

vvcephei committed:
(lower) bound timeout at 0 for ConsumerNetworkClient

vvcephei committed:
fix race condition in consumer coordinator

vvcephei commented:
* rebased to bring in the new scala streams tests (which failed last time).
* fixed a race condition in `ConsumerCoordinator`

vvcephei committed:
async fetchCommittedOffsets

vvcephei commented:
It's been a while since I've recorded status on this. I have a theory that `poll` can strike a balance between blocking and async by being fundamentally async and optionally blocking while it polls for async operations to complete.

Thus, if we need to refresh metadata, and let's say that metadata operation would take 2s and then we have some records to grab, and that would take 3s, then all these scenarios would work as expected:

### completely async:
```
poll(0) => [] // fires off the metadata update requests
sleep(1s)
poll(0) => [] // metadata updates are still pending
sleep(1s)
poll(0) => [] // metadata updates are complete. sends fetch request.
sleep(1s)
poll(0) => [] // fetch request is still pending
sleep(1s)
poll(0) => [] // fetch request is still pending
sleep(1s)
poll(0) => [record1, record2] // response was available
```

### completely sync:
```
poll(5s) => [record1, record2] // sent metadata update request, waited 2s for response, sent fetch request, waited 3s for response
```

### sync, but not long enough to complete the full operation:
```
poll(3s) => [] // sent metadata update request, waited 2s for response, sent fetch request, waited 1s for response
sleep(1s)
poll(3s) => [record1, record2] // waited 1s and got the fetch response.
```

In other words, we make poll completely async but leave in the option to block for responses, since that makes some use cases (especially tests and POCs simpler).

Even with some timeout bigger than 0 but less than infinity, it's ok to block as long as we can and then return an empty list if we didn't finish, whether the thing we didn't finish was a metadata update or a fetch request. This is ok because callers are supposed to *poll* `poll`, aka just keep calling it repeatedly until they get what they want or forever (whichever comes first).

I needed to validate that metadata updates can actually function in an async manner, so (after I added a timeout to the metadata update) I set the metadata update timeout to 0 and put it in a tight loop that would continue until it completed. This is equivalent to the current "block until metadata update completes", but it'll only work if async metadata updates is implemented correctly.

vvcephei commented:
The tests passed on JDK7 and JDK10.

The failure on JDK8 looks like an unrelated failure out of Zookeeper:
```
kafka.api.PlaintextConsumerTest.testMultiConsumerRoundRobinAssignment

org.apache.zookeeper.KeeperException$NodeExistsException: KeeperErrorCode = NodeExists for /config/topics/topic
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:119)
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:51)
	at kafka.zookeeper.AsyncResponse.maybeThrow(ZooKeeperClient.scala:472)
	at kafka.zk.KafkaZkClient.createRecursive(KafkaZkClient.scala:1424)
	at kafka.zk.KafkaZkClient.create$1(KafkaZkClient.scala:262)
	at kafka.zk.KafkaZkClient.setOrCreateEntityConfigs(KafkaZkClient.scala:269)
	at kafka.zk.AdminZkClient.createOrUpdateTopicPartitionAssignmentPathInZK(AdminZkClient.scala:99)
	at kafka.zk.AdminZkClient.createTopic(AdminZkClient.scala:56)
	at kafka.utils.TestUtils$.createTopic(TestUtils.scala:296)
	at kafka.integration.KafkaServerTestHarness.createTopic(KafkaServerTestHarness.scala:123)
	at kafka.api.BaseConsumerTest.setUp(BaseConsumerTest.scala:71)
	at sun.reflect.GeneratedMethodAccessor156.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:114)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:57)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:66)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.GeneratedMethodAccessor10.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy1.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:108)
	at sun.reflect.GeneratedMethodAccessor9.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:146)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:128)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
```

Retest this, please.

vvcephei committed:
deprecate poll and add new methods to complete the proposal

vvcephei commented:
Now that I've validated the approach, I've added the new methods I'm proposing and deprecating the old poll.

vvcephei commented:
One remaining task before this is mergable is to update at least some tests to use either the new `awaitAssignmentMetadata` or the new `poll`, as appropriate.

vvcephei committed:
different semantics for new and old poll methods

vvcephei committed:
remove unnecessary public modifier in interface

vvcephei committed:
Add Duration variants for all the blocking methods

vvcephei commented:
@hachikuji @ConcurrencyPractitioner ,

I've updated this PR with the latest state from KIP-266, as a more concrete example of the proposal. WDYT?

Two things came up for me going through this...

1. The methods Jason proposed to include actually all already used the configs to set a timeout (request.timeout.ms). I was previously unaware of the use of this config in some of Consumer's methods. I'm still in favor of the current approach, but just to point out that a less dramatic change exists... to keep all the old method signatures and just make them respect request.timeout.ms .

2. I've learned that Duration lets you choose durations much greater than Long.MAX_VALUE (which is 25 days). If you try to get millis out of it, though, you get an exception. In contrast, TimeUnit.toMillis will just coerce larger values to Long.MAX_VALUE.  I flirted somewhat with using Duration "all the way down", and you can see some examples of that in KafkaConsumer, but I think it would become unwieldy. Instead, I think a good approach would be to check up front that the duration is "in bounds" and throw an IllegalArgumentException if not.

Alternatives: we could coerce to MAX_VALUE ourselves, or we could use a loop to actually wait for any duration specified.

I'm not sure either of those points are super relevant to the KIP discussion (unless you think so), so I'll leave it to you to bring them up if you're concerned.

I'll be out for a week, so I won't be available to reply to any comments until then.

Note that the tests for this PR won't pass until the CIT agents are updated to java 8 (but you should be able to build it yourself if desired).

vvcephei committed:
fix style

ConcurrencyPractitioner commented:
Hi @vvcephei,

Thanks for setting up the basic methods in the interfaces, but you could just stick with implement poll.
I could take over the rest once you are done.

tedyu commented on clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java:
```diff
@@ -38,156 +39,224 @@
+    /**
+     * @see KafkaConsumer#commitSync()
+     */
+    void commitSync(Duration timeout);
```
Can you move this to after line 95 ?

tedyu commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -543,10 +547,14 @@
 
+    // Since we're universally extracting timeout durations to milliseconds, this is the maximum duration we can accept.
+    // (2^63)-1 milliseconds is 24.86 days.
+    private static final Duration A_MONTH = Duration.ofMillis(Long.MAX_VALUE);
```
24.86 days is not a month :-)

vvcephei replied:
Truth! But I wanted to convey the ballpark to help readers and maintainers understand that unpredicated calls will block for quite a long time, but not "effectively forever", which is the mental mistake I made when I read "max long milliseconds". So now, I have this mental scale: "max long" seconds is effectively forever, with milliseconds it's a month-ish, and with nanoseconds it's about a week.

tedyu commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1107,25 +1122,33 @@ public void assign(Collection<TopicPartition> partitions) {
+
+                client.maybeTriggerWakeup();
+
+                while (!internalUpdateAssignmentMetadataIfNeeded(A_MONTH)) {
```
Shouldn't the time spent in this loop be counted toward ```remaining``` ?

vvcephei replied:
Good eye. But this is the "old" method, whose behavior I'm preserving. So this is a re-expression of the existing behavior, that metadata update will block arbitrarily long and not count against the timeout, followed by the fetch, to which the timeout applies.

guozhangwang commented:
@vvcephei KAFKA-5697 has been fixed via GitHub Pull Request #4930, so I'm wondering if it is more for KIP-266 itself?

ConcurrencyPractitioner commented:
@vvcephei  I noticed that in KIP-266, we have several methods which does not seem to have a problem with blocking indefinitely, e.g. offsetsByTimes, beginningOffsets, endOffsets, so do we need to change those methods?

tedyu commented:
retest this please

vvcephei commented:
@ConcurrencyPractitioner Sure thing; I'll pull the extra stuff back out. it wasn't as trivial as I thought it would be, evidenced by the fact that the tests broke, so I agree that poll alone is a big enough piece for one PR.

vvcephei commented:
@tedyu Thanks for the reivew! The test failures were related. Something in the "extra methods" I popped in caused the test failures, but I'm about to take them back out, so we should be green again.

vvcephei committed:
roll back extra implementations

vvcephei commented:
Ok, bringing this PR back to life. I've pulled out the "extra" methods I'd added. I expect the tests to pass.

Next up, I need to write a few tests for the deprecated method and migrate the rest of the code base and tests away from it.

@hachikuji should I *just* migrate the `clients` code base and leave the rest of kafka alone? Or migrate the whole project? I'm not sure what the norm with this project is.

vvcephei committed:
fix style

guozhangwang commented:
@vvcephei thanks for the updated PR, please ping me and @hachikuji whenever it is ready to be reviewed. Please do the following to help reviewers better understand the PR diffs:

1. For the old API implementation, if we HAVE TO make any changes to it (personally I'd prefer to not make any changes to keep its scope small, but I'll leave it to you), state clearly in the description what are those and making sure they do not change any semantics of the old API.

2. For the new API implementation, describe its semantics that given a (possibly zero) timeout value, what will happen if a) it need to send metadata, b) it needs to (re-)join group, c) its metadata is fresh and group is stable, and there is inflight fetch request, d) its metadata is fresh and group is stable, and there is NO inflight fetch request.

3. Also it would help to add a few more bullet points on the actual changes in KafkaConsumer and AbstractCoordinator to help better understanding the diff details.

vvcephei committed:
transition Clients tests and comments to new poll

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java:
```diff
@@ -38,156 +39,162 @@
     /**
      * @see KafkaConsumer#assignment()
      */
-    public Set<TopicPartition> assignment();
```
`public` is redundant in an interface definition.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerInterceptor.java:
```diff
@@ -35,14 +35,16 @@
  * <p>
- * ConsumerInterceptor callbacks are called from the same thread that invokes {@link org.apache.kafka.clients.consumer.KafkaConsumer#poll(long)}.
+ * ConsumerInterceptor callbacks are called from the same thread that invokes
+ * {@link org.apache.kafka.clients.consumer.KafkaConsumer#poll(java.time.Duration)}.
```
The first of many comments that need to change from `poll(long)` to `poll(Duration)`, since we shouldn't reference the deprecated version in javadoc.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1107,25 +1117,33 @@ public void assign(Collection<TopicPartition> partitions) {
-            long start = time.milliseconds();
             long remaining = timeout;
             do {
-                Map<TopicPartition, List<ConsumerRecord<K, V>>> records = pollOnce(remaining);
```
the old `pollOnce` has been replaced by `internalUpdateAssignmentMetadataIfNeeded` + `pollForFetches`.

We preserve the prior semantics of blocking forever on metadata updates without counting that time against the timeout.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1107,25 +1117,33 @@ public void assign(Collection<TopicPartition> partitions) {
+
+                while (!internalUpdateAssignmentMetadataIfNeeded(Duration.ofMillis(Long.MAX_VALUE))) {
+                    log.warn("Still waiting for metadata");
+                }
```
Since `internalUpdateAssignmentMetadataIfNeeded` has a maximum timeout of `max long` ms (24.85 days), we call it in a loop. I doubt that the client will block here for almost a month without anyone noticing, but it's better to just stick to the current behavior, which is an indefinite block.

vvcephei committed:
wip

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1107,25 +1117,32 @@ public void assign(Collection<TopicPartition> partitions) {
+            final long start = time.milliseconds();
             long remaining = timeout;
             do {
-                Map<TopicPartition, List<ConsumerRecord<K, V>>> records = pollOnce(remaining);
```
the internal method `pollOnce` has been split into `internalUpdateAssignmentMetadataIfNeeded` and `pollForFetches` so that both the new and old methods can provide their semantics without having to maintain full copies of the nontrivial internal code.

vvcephei committed:
document upper bound on poll duration

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1136,56 +1153,129 @@ public void assign(Collection<TopicPartition> partitions) {
+     * @throws java.lang.IllegalArgumentException if the timeout value is negative
+     * @throws java.lang.IllegalStateException if the consumer is not subscribed to any topics or manually assigned any
+     *             partitions to consume from
+     * @throws java.lang.ArithmeticException if the timeout is greater than {@link Long#MAX_VALUE} milliseconds.
```
Note this new exception.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1136,56 +1153,129 @@ public void assign(Collection<TopicPartition> partitions) {
+        }
+
+        return updateFetchPositions(remainingMsAtLeastZero(startMs, timeout));
+    }
```
This algorithm comes from the first part of `pollOnce`.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1426,7 +1516,9 @@ public long position(TopicPartition partition) {
-                updateFetchPositions();
+                while (!updateFetchPositions(Long.MAX_VALUE)) {
+                    log.warn("Still updating fetch positions");
+                }
```
preserving the old blocking behavior, even though the internal method now takes a timeout.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1457,7 +1549,13 @@ public long position(TopicPartition partition) {
+                    Collections.singleton(partition),
+                    Long.MAX_VALUE
+                );
+            }
```
preserving the old blocking behavior, even though the internal method now takes a timeout.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1136,56 +1153,129 @@ public void assign(Collection<TopicPartition> partitions) {
         // We do not want to be stuck blocking in poll if we are missing some positions
         // since the offset lookup may be backing off after a failure
-        if (!hasAllFetchPositions && pollTimeout > retryBackoffMs)
+        if (!subscriptions.hasAllFetchPositions() && pollTimeout > retryBackoffMs)
```
We used to call `updateFetchPositions` and save the result in `hasAllFetchPositions` earlier in this method. Now, we call `updateFetchPositions` in the metadata update method before entering this method, so we equivalently check directly if we have all fetch positions.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1772,18 +1873,17 @@ private void close(long timeoutMs, boolean swallowException) {
-    private boolean updateFetchPositions() {
-        if (subscriptions.hasAllFetchPositions())
-            return true;
+    private boolean updateFetchPositions(final long timeoutMs) {
```
This method used to block indefinitely, so we have to add a timeout parameter. Callers need to know whether it timed out or not, and they can easily check directly if all partitions have an assigned position, so we're using the boolean return value to indicate whether the operation completed or timed out.

We could throw and catch timeout exceptions instead, but I felt it was cleaner to return instead, and we can also avoid a context switch this way.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1772,18 +1873,17 @@ private void close(long timeoutMs, boolean swallowException) {
         // a consumer with manually assigned partitions can avoid a coordinator dependence
         // by always ensuring that assigned partitions have an initial position.
-        coordinator.refreshCommittedOffsetsIfNeeded();
+        if (!coordinator.refreshCommittedOffsetsIfNeeded(timeoutMs)) return false;
```
If this call times out, then `updateFetchPositions` as a whole times out.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1835,4 +1935,15 @@ private void throwIfNoAssignorsConfigured() {
+
+    private long remainingMsAtLeastZero(final long startMs, final Duration timeout) {
+        return Math.max(0, timeout.minusMillis(time.milliseconds() - startMs).toMillis());
+    }
```
Convenience methods to decrement the remaining time for timeouts. We generally don't want the timeout to become negative, since internal calls rightly expect not to get negative timeouts as parameters, and at the least we're firing off asynchronous requests, which algorithmically is instantaneous, but in reality takes non-zero ms.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java:
```diff
@@ -146,8 +147,14 @@ public synchronized void unsubscribe() {
+    @Deprecated
     @Override
     public synchronized ConsumerRecords<K, V> poll(long timeout) {
+        return poll(Duration.ZERO);
```
The timeout is ignored in the mock consumer, so zero is as good as any value.

vvcephei committed:
wip

vvcephei committed:
wip

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -198,34 +198,30 @@ protected abstract void onJoinComplete(int generation,
      */
-    protected synchronized boolean ensureCoordinatorReady(long startTimeMs, long timeoutMs) {
-        long remainingMs = timeoutMs;
+    public synchronized boolean ensureCoordinatorReady(final long timeoutMs) {
```
While adding timeouts to internal methods, I encountered two pre-existing internal timeout styles. Some took a `startTime` parameter, and others call `time.milliseconds` to determine the start time. I assume the former style is a holdover from before the `Time` interface was available. I went ahead and standardized on `Time` when I needed to touch the method anyway.

hachikuji replied:
It was actually because of the point mentioned before about avoiding unnecessary system calls. It's just kind of annoying when you have to pass it down deep into the stack, so we may not have followed the pattern consistently.

Also, why change this to public?

vvcephei replied:
That makes sense. I was thinking that you'd have the same number of calls (either on the caller or on the callee's side), but that's the case only if you're trying to be super accurate. I think the deadline style you proposed would eliminate this problem.

I guess an equivalent to the deadline approach is to pass in (start,timeout) everywhere, but instead of re-computing start and timeout for each inner call, just using the start and timeout of the outermost request. This approach didn't occur to me before. Deadline is simpler still, though, since it's a scalar value and contains the same information.

vvcephei replied:
Regarding the visibility change: There was another method `public synchronized void ensureCoordinatorReady()` that delegated to this one. It needed to be public because the WorkerCoordinator in Streams wants to expose it for a test. I got rid of that one to force all callers to specify a timeout.

I can fix it by adding a package-private delegating method in WorkerCoordinator to expose the method for its test.

vvcephei committed:
wip

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -309,11 +298,29 @@ protected synchronized long timeToNextHeartbeat(long now) {
     public void ensureActiveGroup() {
+        while (!ensureActiveGroup(Long.MAX_VALUE)) {
+            log.warn("still waiting to ensure active group");
+        }
```
again, preserving the old indefinite block.

vvcephei committed:
wip

vvcephei committed:
wip

vvcephei committed:
wip

vvcephei committed:
migrate startTime style

vvcephei committed:
wip

vvcephei committed:
wip

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -440,55 +455,88 @@ public boolean needRejoin() {
 
+    private Set<TopicPartition> pendingCommittedOffsetRequest = null;
+    private RequestFuture<Map<TopicPartition, OffsetAndMetadata>> pendingCommittedOffsetResponse = null;
+
```
This might be the sketchiest thing I've done. Previously, we wouldn't keep pending requests around, but just start a new one on every call. This was ok because we would always wait for completion before returning.

Now, we can return early if we time out, so it behooves us to implement async request semantics as we have elsewhere (notably in fetch).

The trick is that we also have to save the request along with the response future, since after a timeout subsequent calls may be for a different set of partitions. It wouldn't be right to return offsets for a different set of partitions than the ones requested, so in that case, we just throw away the previous future and save the new one.

This is not thread-safe because I think it can only be called in one thread.

hachikuji replied:
I think the basic idea makes sense and is what I was expecting. It might not feel too elegant, but I think a simple approach is best initially. An interesting point to consider is what would happen if an offset fetch is in-flight while a rebalance is in progress. When it returns, the offsets may be stale. I am wondering if it makes sense to fence the response using the group generation. In other words, we record the generation at the time of the request and then verify when we receive the response that it matches the current group generation.

vvcephei replied:
That sounds safer. I'll go ahead and wrap the three things (request, future, and request-time generation) in a small inner class, just to keep things tidy.

vvcephei commented on clients/src/test/java/org/apache/kafka/clients/consumer/DeprecatedMethodsKafkaConsumerTest.java:
```diff
@@ -0,0 +1,1736 @@
+import static org.junit.Assert.fail;
+
+@SuppressWarnings({"deprecation", "SameParameterValue"})
+public class DeprecatedMethodsKafkaConsumerTest {
```
We need to keep testing the methods we're deprecating, so I copied the whole `KafkaConsumerTest`. Then, I updated `KafkaConsumerTest` to use the new `poll(Duration)`. Once we get rid of the deprecated methods in a future version, we can just delete this test.

hachikuji replied:
There has to be a better way to do this. For example, maybe we can extend `KafkaConsumerTest`?

vvcephei replied:
I see where you're coming from, but I unironically think this is a great way to do this.

In general, we just want to make sure that we continue testing the same conditions we currently are. Capturing a snapshot of our current tests is a pretty sure way to do this.

I considered deleting everything that isn't directly relevant to the `poll` methods, but then again, @ConcurrencyPractitioner is going deprecate a bunch more methods, so I think he'll just have to add most of it back.

If we subclass `KafkaConsumerTest`, then you have the test harness separated from the tests, and when you modify `KafkaConsumerTest` in the future, you have to worry about also modifying these tests.

We could also just duplicate the relevant tests in `KafkaConsumerTest`, but that would make that file almost twice as long, therefore harder to read. Also, at some point, we'll want to remove the deprecated methods. When that happens, you can just delete this whole file, as opposed to having to go through `KafkaConsumerTest` and disentangle the deprecated-method and non-deprecated-method tests.

vvcephei commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -385,7 +392,9 @@ public void verifyHeartbeatSentWhenFetchedDataReady() throws Exception {
-        consumer.poll(0);
+        consumer.internalUpdateAssignmentMetadataIfNeeded(Duration.ZERO);
+        consumer.internalUpdateAssignmentMetadataIfNeeded(Duration.ZERO);
+        consumer.poll(Duration.ZERO);
```
We need to issue the poll request because the mock is expecting it, but we also need the double-metadata call to initialize the metadata beforehand, otherwise the internal code won't issue the fetch request.

vvcephei commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -425,7 +434,7 @@ public void verifyNoCoordinatorLookupForManualAssignmentWithSeek() {
         client.prepareResponse(fetchResponse(tp0, 50L, 5));
 
-        ConsumerRecords<String, String> records = consumer.poll(5);
+        ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(1));
```
time never actually advances because we use `MockTime(0)`, so any timeout at all will block indefinitely and keep polling until we get a response. This only works in tests that expect a response, otherwise it would loop forever.

vvcephei commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -353,7 +354,12 @@ public void verifyHeartbeatSent() throws Exception {
+        // 2. the second one gets the response we already queued up
+        consumer.internalUpdateAssignmentMetadataIfNeeded(Duration.ZERO);
+        consumer.internalUpdateAssignmentMetadataIfNeeded(Duration.ZERO);
+
```
I hope that comment makes sense. Alternatively, we could use one call with a non-zero timeout (such as 1ms), but I liked the double async call, since it seems to be a tighter bound on the behavior.

vvcephei commented on clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java:
```diff
@@ -34,6 +35,32 @@
+
+    @SuppressWarnings("deprecation")
+    @Test
+    public void testSimpleMockDeprecated() {
```
Again, I copied the original test so we can continue to test the deprecated method.

vvcephei commented on clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java:
```diff
@@ -34,6 +35,32 @@
 
     @Test
     public void testSimpleMock() {
```
I modified this test to use the new `poll(Duration)`.

vvcephei commented on clients/src/test/java/org/apache/kafka/common/security/authenticator/ClientAuthenticationFailureTest.java:
```diff
@@ -85,12 +86,12 @@ public void testConsumerWithInvalidCredentials() {
         try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props, deserializer, deserializer)) {
             consumer.subscribe(Arrays.asList(topic));
-            consumer.poll(100);
+            consumer.poll(Duration.ofSeconds(10));
```
Time actually does pass during this test, so we need the timeout to account for the metadata update. On my machine, this test takes ~800ms to complete, so I figure 10s should be enough elbow room.

vvcephei commented on clients/src/test/java/org/apache/kafka/common/security/authenticator/ClientAuthenticationFailureTest.java:
```diff
@@ -85,12 +86,12 @@ public void testConsumerWithInvalidCredentials() {
             // OK
         } catch (Exception e) {
-            fail("Expected only an authentication error, but another error occurred: " + e.getMessage());
+            throw new AssertionError("Expected only an authentication error, but another error occurred.", e);
```
This way, we can get the stack trace in the output.

vvcephei commented on streams/src/test/java/org/apache/kafka/test/MockRestoreConsumer.java:
```diff
@@ -85,6 +85,7 @@ public synchronized void assign(Collection<TopicPartition> partitions) {
         super.assign(partitions);
     }
 
+    @Deprecated
```
This isn't in the `:clients` project, but it is the only other subclass of `Consumer`. I want to make sure we don't mask the deprecation.

vvcephei reviewed (commented):
I've made a pass over this (quite large) pr to make sure it's ready for review and to add some explanatory remarks.

Side note: over 1k LOC in the diff are simply copying the consumer test so we can test both the new and deprecated methods.

vvcephei commented:
@hachikuji and @guozhangwang , this PR is ready for review.

@guozhangwang, I tried to address your requests in the PR description. Please let me know if my explanation is not adequate.

vvcephei commented:
@ConcurrencyPractitioner I've finally finished this PR. WDYT?

ConcurrencyPractitioner commented:
Hi @vvcephei, it looks good to me. I don't really know if the changes to ```refreshCommittedOffsetsIdNeeded``` are practical as you stated, but other than that. I couldn't find any problems.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -180,7 +181,8 @@
  * The recommended way to handle these cases is to move message processing to another thread, which allows
- * the consumer to continue calling {@link #poll(long) poll} while the processor is still working. Some care must be taken
+ * the consumer to continue calling {@link #poll(Duration) poll} while the processor is still working. Some care must be
+ * taken
```
nit: reflow?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1136,56 +1153,128 @@ public void assign(Collection<TopicPartition> partitions) {
+     * Visible for testing
+     */
+    @SuppressWarnings("BooleanMethodIsAlwaysInverted") // because false => timed out, in which case we return early or throw.
+    boolean internalUpdateAssignmentMetadataIfNeeded(final Duration timeout) {
```
nit: the "internal" prefix seems like overkill

vvcephei replied:
I called it "internal" because this method isn't protected by the light lock. It seemed like a risk that someone might want to expose this method in the future, and I wanted them to think twice about just making it public.

Alternatively, I can go ahead and (re)acquire the lock in here, or just change the name to "updateAssignmentMetadataIfNeeded" and leave the implementation as-is.

I'm happy to do whatever you prefer.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1096,9 +1099,16 @@ public void assign(Collection<TopicPartition> partitions) {
     @Override
-    public ConsumerRecords<K, V> poll(long timeout) {
+    public ConsumerRecords<K, V> poll(final long timeout) {
+        // note that this method can't defer to its replacement because its semantics are different wrt waiting for
```
It seems like we ought to be able to handle both overloads with a single method takes a timeout and a boolean flag indicating whether to wait forever for metadata update.

vvcephei replied:
yeah, we can do that, if you prefer.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1772,18 +1872,17 @@ private void close(long timeoutMs, boolean swallowException) {
-    private boolean updateFetchPositions() {
-        if (subscriptions.hasAllFetchPositions())
-            return true;
+    private boolean updateFetchPositions(final long timeoutMs) {
```
This is a bit of a meta comment. We have structured all of these internal APIs to use a timeout. It makes sense in isolation, but when you have a sequence of operations, it's annoying to repeatedly recompute the remaining time. I wonder what it would look like to structure them in terms of deadlines instead. So we just compute a deadline at the start of the call in `KafkaConsumer` and carry it through each operation.

The other thing is that we seem to have more calls to system time than we had before. For example, there are three in `internalUpdateAssignmentMetadataIfNeeded` which would hit every call to poll(). It may not seem like a big deal, but we have actually found it to be a performance bottleneck in some cases, which is why we do seemingly silly optimizations like passing around the current time. Consider a case where someone configures `max.poll.records` to 1, then every record return to the user will be accompanied by a bunch of system calls. It would be nice to reduce these calls at least for the common path where we have all our offsets and we are just awaiting fetches.

vvcephei replied:
The deadline thing would be more elegant, but I'm concerned that, carelessly applied, it would spider through the clients code base; I think I'd wind up changing much more than I strictly have to. I'm not opposed, and I'll go ahead and give it a crack to see what it looks like, just wanted to point this out.

vvcephei replied:
About the extra calls to system time... Yeah, I did that deliberately because it made the code simpler, but I didn't consider that the call itself could become a bottleneck, which is completely plausible. I'll switch it back to the prior strategy, of only checking the time when it really matters.

Applying the deadline-instead-of-timeout approach would also save one call to system time at the start of every method to get the "startTime".

hachikuji reviewed (commented):
Thanks for the patch! I'm still making my way through the code, but thought I'd post one high level comment to see what you think.

vvcephei committed:
reflow

vvcephei committed:
consolidate poll logic

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1096,96 +1097,156 @@ public void assign(Collection<TopicPartition> partitions) {
+        return poll(timeout, true);
+    }
+
+    private ConsumerRecords<K, V> poll(final Duration timeout, final boolean includeMetadataInTimeout) {
```
@hachikuji I've consolidated this logic as requested. Note that the timing here is slightly different for the deprecated use case. On this latest pass though the code, I noticed that the original `poll(long)` actually counts time spent in metadata *after the first loop*.

This would only matter if the timeout exceeds the "time to next poll" in the underlying consumer, since that's the only case in which we would have more than one time through the loop.

I think it's not that important, but figured it was worth pointing out.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1096,96 +1097,156 @@ public void assign(Collection<TopicPartition> partitions) {
         // after the long poll, we should check whether the group needs to rebalance
         // prior to returning data so that the group can stabilize faster
-        if (coordinator.needRejoin())
+        if (coordinator.needRejoin()) {
```
nit: avoiding the braces has been the convention in kafka. I was never too fond of the convention and I'm not strict about enforcing it for new code, but it's a little annoying and distracting to change the old code.

vvcephei replied:
ok, sorry; it happened in a few places when I applied auto-formatting to the methods I was touching. I'll make a pass through and revert formatting changes.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1096,96 +1097,156 @@ public void assign(Collection<TopicPartition> partitions) {
                     return this.interceptors.onConsume(new ConsumerRecords<>(records));
                 }
+                final long fetchEnd = time.milliseconds();
+                remainingMs = Math.max(0L, remainingMs - (fetchEnd - fetchStart));
```
We don't need super precise timing control for any of these APIs, but the advantage of the previous approach is that the timeout is effectively a deadline and you won't accrue rounding errors while updating the time remaining. I'm not sure what the reason to change the logic was.

vvcephei replied:
Yeah... This is a result of double-refactoring. I got away from the "elapsed" approach when I introduced the "remainingTimeAtLeastZero" method, and then I made a recent change to avoid calling system time so much. I would have gone back to the original time-tracking approach, but I'd forgotten about it.

You are right, tracking elapsed time doesn't rack up rounding errors. I'll add another patch momentarily using this approach throughout my diff instead of "remainingTime".

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1096,96 +1097,156 @@ public void assign(Collection<TopicPartition> partitions) {
 
-        long startMs = time.milliseconds();
-        coordinator.poll(startMs, timeout);
+        return updateFetchPositions(remainingMsAtLeastZero(startMs, timeout));
```
Since `Coordinator.poll()` can block, we probably have to update `startMs` here.

vvcephei replied:
That's why we're calling `remainingMsAtLeastZero` for a new timeout here. Although I might be about to change this to use elapsed-time instead.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1096,96 +1097,156 @@ public void assign(Collection<TopicPartition> partitions) {
         // We do not want to be stuck blocking in poll if we are missing some positions
         // since the offset lookup may be backing off after a failure
-        if (!hasAllFetchPositions && pollTimeout > retryBackoffMs)
+        if (!subscriptions.hasAllFetchPositions() && pollTimeout > retryBackoffMs) {
```
The previous code tried to avoid unnecessary passes over the assigned partitions since the set can get fairly large. I'm not sure how beneficial this optimization was. Maybe it's fine to lose it. Nevertheless, since `poll` can be called quite tightly, it would be nice to avoid redundant work.

vvcephei replied:
hmm. I'm not sure what a good way to do this is... `hasAllFetchPositions` was the result of `updateAllFetchPositions` before (which changed because the result now indicates whether it timed out or not). But `updateAllFetchPositions` is not longer called in this scope, so we can't consider keeping its result around.

I could cache the "hasAllFetchPositions" fact on a field in KafkaConsumer, but it seems a little risky. I *think* it would be fine... WDYT?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -198,34 +198,30 @@ protected abstract void onJoinComplete(int generation,
-     * @param startTimeMs Current time in milliseconds
-     * @param timeoutMs Maximum time to wait to discover the coordinator
+     *
+     * @param timeoutMs   Maximum time to wait to discover the coordinator
```
nit: was the spacing intentional?

vvcephei replied:
Sort of; my formatter is set to align the param descriptions. But now that the other param is gone, it's aligned to a ghost...

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -265,11 +257,8 @@ private synchronized void clearFindCoordinatorFuture() {
      * @return true if it should, false otherwise
      */
     protected synchronized boolean needRejoin() {
```
With the change here, maybe we should consider renaming this. Perhaps `awaitingRejoin`?

vvcephei replied:
When I renamed this method, I realized it's used way more widely than I'd thought. I looked at the other usages, and they all *look* safe...

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -347,8 +354,23 @@ private void closeHeartbeatThread() {
     void joinGroupIfNeeded() {
-        while (needRejoin() || rejoinIncomplete()) {
-            ensureCoordinatorReady();
+        while (!joinGroupIfNeeded(Long.MAX_VALUE)) {
```
As before, the old code tried to optimize for the common case where the consumer was joined.

vvcephei replied:
In this case, the new and old code does the same thing. We're just delegating to the new variant that has a timeout. Although I just realized that this method is only called in tests now, so I'll just inline it.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -268,30 +268,39 @@ protected void onJoinComplete(int generation,
+     * @return true iff the operation succeeded
      */
-    public void poll(long now, long remainingMs) {
+    public boolean poll(final long timeoutMs) {
```
Same comment as before. We can skip the time system calls by passing the current time since the common case is that no work needs to be done here.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -440,55 +455,88 @@ public boolean needRejoin() {
+            pendingCommittedOffsetResponse = null;
+        }
+
+        final long startMs = time.milliseconds();
```
nit: move below the shortcut return below.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -440,55 +455,88 @@ public boolean needRejoin() {
     }
 
+    private Set<TopicPartition> pendingCommittedOffsetRequest = null;
+    private RequestFuture<Map<TopicPartition, OffsetAndMetadata>> pendingCommittedOffsetResponse = null;
```
Can we move these to the top with the rest of the fields? Maybe we can choose clearer names?

vvcephei replied:
yep. I'll also wrap them in an inner class, since we're going to add generation too.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:
```diff
@@ -165,8 +165,22 @@ public boolean awaitMetadataUpdate(long timeout) {
      */
     public void ensureFreshMetadata() {
-        if (this.metadata.updateRequested() || this.metadata.timeToNextUpdate(time.milliseconds()) == 0)
+        if (this.metadata.updateRequested() || this.metadata.timeToNextUpdate(time.milliseconds()) == 0) {
```
Could we replace this with `ensureFreshMetadata(Long.MAX_VALUE)`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:
```diff
@@ -165,8 +165,22 @@ public boolean awaitMetadataUpdate(long timeout) {
+
+    boolean ensureFreshMetadata(final long timeout) {
+        if (timeout < 0) {
+            return false;
```
Could we just raise `IllegalArgumentException` as we do in most other cases?

hachikuji reviewed (commented):
Thanks for the patch. Left a few comments.

vvcephei committed:
be stingier with time.milliseconds()

vvcephei committed:
restrict method visibility

vvcephei committed:
rename needRejoin

vvcephei committed:
inline unneeded method

vvcephei committed:
revert spacing

vvcephei committed:
move shortcut to top of method

vvcephei committed:
add generation to pending offset request

vvcephei committed:
get rid of unnecessary ensureFreshMetadata variant

vvcephei committed:
remove unnecessary precondition

vvcephei committed:
CR comments

vvcephei commented:
Allrighty, @hachikuji,

Thanks so much for the thorough review. I think I've addressed all your concerns.

Regarding the timeout-vs-deadline discussion. I've refactored all my timeout handling in terms of elapsed time and minimized the system calls. I think this is good enough for now. I would agree with switching over to deadlines in the future, but it feels too big to lump in with this work.

Thanks,
-John

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1136,56 +1201,61 @@ public void assign(Collection<TopicPartition> partitions) {
+
+        // NOTE: the use of cachedSubscriptionHashAllFetchPositions means we MUST call
+        // updateAssignmentMetadataIfNeeded before this method.
+        if (!cachedSubscriptionHashAllFetchPositions && pollTimeout > retryBackoffMs) {
```
We might also consider moving this into `SubscriptionState`? We can invalidate the value whenever the assignment changes or an offset is reset. That would also give us a shortcut for `missingFetchPositions()`.

vvcephei replied:
I think we'd better defer this for later.

hachikuji replied:
Sounds good.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -264,12 +262,9 @@ private synchronized void clearFindCoordinatorFuture() {
-
-    private synchronized boolean rejoinIncomplete() {
-        return joinFuture != null;
+    protected synchronized boolean rejoinNeededOrPending() {
```
We should probably update the javadoc

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -86,6 +87,31 @@
+        }
+
+        private boolean sameRequest(final Set<TopicPartition> currentRequest, final Generation currentGeneration) {
+            return currentGeneration.generationId == generation.generationId
```
Maybe we can implement `equals` for `Generation`?

vvcephei replied:
I considered it, but wanted to make sure this was in fact the comparison we should be doing here. I guess it is.

hachikuji commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java:
```diff
@@ -103,6 +103,12 @@ public String protocolType() {
 
+    // expose for tests
+    @Override
+    protected synchronized boolean ensureCoordinatorReady(final long timeoutMs) {
```
Do we need this? Seems like the same signature as the parent.

vvcephei replied:
It's weird, but doing this makes the method accessible in the `WorkerCoordinatorTest`, which it isn't otherwise because the test is in a different package from `AbstractCoordinator`. `WorkerCoordinator` can see the superclass's method because it is a subclass.

hachikuji replied:
Makes sense.

hachikuji commented on core/src/test/scala/unit/kafka/zk/ZooKeeperTestHarness.scala:
```diff
@@ -121,6 +121,10 @@ object ZooKeeperTestHarness {
-    assertTrue(s"Found unexpected threads during $context, allThreads=$threads", noUnexpected)
+    assertTrue(
+      s"Found unexpected threads during $context, allThreads=$threads, " +
+        s"unexpected=${threads.filterNot(t => unexpectedThreadNames.forall(s => !t.contains(s)))}",
```
👍

hachikuji reviewed (commented):
Thanks for the updates. Just a few more comments.

vvcephei committed:
style

vvcephei committed:
oops

vvcephei committed:
replace comprehensive deprecated-method test with simple code-path test

vvcephei committed:
javadoc update

vvcephei committed:
move generation equality to Generation

vvcephei committed:
account for null generation

vvcephei commented:
Ok, @hachikuji , I think this PR is gtg now, supposing the tests pass.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -349,7 +394,7 @@ public long timeToNextPoll(long now) {
         // update metadata (if needed) and keep track of the metadata used for assignment so that
         // we can check after rebalance completion whether anything has changed
-        client.ensureFreshMetadata();
+        if (!client.ensureFreshMetadata(Long.MAX_VALUE)) throw new TimeoutException();
```
I guess it's tough to avoid blocking here. All of this would be considerably easier if we moved the rebalance to the background thread. Another improvement for another time.

hachikuji reviewed (approved):
LGTM. Thanks for the patch

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -86,6 +87,32 @@
+    private PendingCommittedOffsetRequest pendingCommittedOffsetRequest = null;
+
+    private static class PendingCommittedOffsetRequest {
+        private final Set<TopicPartition> request;
```
The names here are a bit misleading, maybe we can fix them in https://github.com/apache/kafka/pull/5087? cc @hachikuji . I'm thinking `requestedPartitions` and `requestedGeneration`?

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -301,18 +341,23 @@ public void poll(long now, long remainingMs) {
+        return true;
+    }
+
+    private long remainingTimeAtLeastZero(final long timeoutMs, final long elapsed) {
```
We have three `remainingTimeAtLeastZero` functions, in AbstractCoordinator, ConsumerCoordinator and KafkaConsumer. Is it intentional? If not we could leave just one to avoid unintentional code divergence in the future. cc @vvcephei

guozhangwang commented:
Cherry-picked to 2.0 since it is an important bug fix to improve operational experience.
